modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/FieldSelectModalContent.tsx

```diff
@@ -0,0 +1,364 @@
+	const [expandedKeys, setExpandedKeys] = useState<Array<React.Key>>([]);
+	const [searchCounter, setSearchCounter] = useState<number>(0);
+
+	useEffect(() => {
+		getFields(fdsView).then((fields) => {
+			if (fields) {
+				const [initialSelectedKeys, updatedFields] = initializeFields({
+					fields,
+					selectedFields,
+				});
+
+				setSelectedKeys(initialSelectedKeys);
+
+				setInitialFields(updatedFields);
+				setFields(updatedFields);
+			}
+		});
+	}, [selectedFields, fdsView]);
+
+	const onSearch = (query: string) => {
+		setQuery(query);
+
+		const {counter, filteredItems, filteredKeys} = applyFilter({
+			fields: initialFields ?? [],
+			query,
+		});
+
+		setFields(filteredItems);
+		setExpandedKeys(filteredKeys);
+		setSearchCounter(counter);
+	};
+
+	return (
+		<>
+			<ClayModal.Header>
+				{sub(
+					Liferay.Language.get('select-x'),
+					Liferay.Language.get('field')
+				)}
+			</ClayModal.Header>
+
+			<ClayModal.Body className="pt-0 px-0">
+				{fields === null ? (
+					<ClayLoadingIndicator />
+				) : (
+					<>
+						<ClayManagementToolbar>
+							<ClayManagementToolbar.Search>
+								<Search onSearch={onSearch} query={query} />
+							</ClayManagementToolbar.Search>
+						</ClayManagementToolbar>
+
+						{query && (
+							<ClayResultsBar>
+								<ClayResultsBar.Item expand>
+									<span className="component-text text-truncate-inline">
+										<span className="text-truncate">
+											{sub(
+												searchCounter === 1
+													? Liferay.Language.get(
+															'x-result-for-x'
+													  )
+													: Liferay.Language.get(
+															'x-results-for-x'
+													  ),
+												searchCounter,
+												query
+											)}
+										</span>
+									</span>
+								</ClayResultsBar.Item>
+
+								<ClayResultsBar.Item>
+									<ClayButton
+										className="component-link tbar-link"
+										displayType="unstyled"
+										onClick={() => {
+											setQuery('');
+											setFields(initialFields);
+										}}
+									>
+										{Liferay.Language.get('clear')}
+									</ClayButton>
+								</ClayResultsBar.Item>
+							</ClayResultsBar>
+						)}
+
+						<div className="container-fluid container-fluid-max-xl px-4 py-2">
+							<TreeView
+								className="bg-light"
+								expandedKeys={new Set(expandedKeys)}
+								items={fields}
+								nestedKey="children"
+								onExpandedChange={(keys) => {
+									setExpandedKeys(Array.from(keys));
+								}}
+								onItemsChange={(items) =>
+									setInitialFields(
+										items as Array<IFieldTreeItem>
+									)
+								}
+								onSelectionChange={setSelectedKeys}
+								selectedKeys={selectedKeys}
+								selectionMode={selectionMode}
+								showExpanderOnHover={false}
+							>
+								{({children, label, query}: IFieldTreeItem) => (
+									<TreeView.Item>
+										<TreeView.ItemStack>
+											<ClayCheckbox checked>
```

Comment:
The modal UI does not match mockups:
- there should be no checkboxes
- we should only allow selection of leaf items

We can do this in a [separate story.](https://liferay.atlassian.net/browse/LPD-6702)